### DISCUSSION TITLE
is auth=PSK possible through softhsm2

### DISCUSSION THREAD

**Discussion by satishkrishna687** [strongswan/strongswan#1516] [Q&A] [answered]
So, I have been trying to store a psk as "secret key" inside softhsm token. But when I mention the handle the "secret key" inside it doesn't load. I checked the logs It only searches for private key for the specificied handle
secrets {
token-soft{
handle=handle_no
Slot= slot_no
 }
}
    Is there a way I can retrieve the "secret key"  to use to authenticate it as psk

I generated the secret key using openssl
openssl rand -hex 32 > mypskfile
and I inserted it into token using 
Pkcs11-tool --write-object

**Comment by tobiasbrunner** [comment, marked as answer]
No, you'd have to write your own plugin that implements something like this, or possibly extend the _pkcs11_ plugin.

**Reply by satishkrishna687** [reply]
Thank you @tobiasbrunner 😁

**Reply by satishkrishna687** [reply]
@tobiasbrunner is the above constraint limited to only softhsm or would a actual hsm work, Basically I want to have a pre shared key on a HSM(any HSM will do) and establish a IPsec connection between a host to host with it. Is there a possibility that I can do this with strongswan, if not are there any alternative of  implementing the above requirement.

**Reply by tobiasbrunner** [reply]
No, there currently isn't. Without custom or modified plugins PSKs can only be stored in text files and databases (there is also no concept of associating a handle/slot etc. with a PSK in swanctl.conf).

**Reply by satishkrishna687** [reply]
Coming away from HSM's, So I stored my psk inside a text file, but I don't see any options in the secrets section to retrieve my text file. The only option that seems to supports psk is through Ike<suffix> part of secrets section.

**Reply by tobiasbrunner** [reply]
With "text files" I meant swanctl.conf (i.e. the `secrets` section). You could include secrets from separate files but you can't load PSKs from arbitrary text files.
